angular-holiday-planner/src/app/components/login/login.component.ts

```diff
@@ -0,0 +1,54 @@
+import { Component } from '@angular/core';
+import {FormControl, FormGroup, Validators} from "@angular/forms";
+import {IUser} from "../../models/user";
+import {fromPromise} from "rxjs/internal/observable/innerFrom";
+import {catchError, first, switchMap} from "rxjs";
+import {AuthService} from "../../services/auth.service";
+import {UserCredential} from "@angular/fire/auth";
+import {Router} from "@angular/router";
+
+@Component({
+  selector: 'app-login',
+  templateUrl: './login.component.html',
+  styleUrls: ['./login.component.scss']
+})
+export class LoginComponent {
+
+  constructor(private auth: AuthService, private router: Router) {}
+
+  loginForm = new FormGroup({
+    email: new FormControl('', [Validators.required]),
+    password: new FormControl('', [Validators.required])
+  });
+
+  get email() {
+    return this.loginForm.get("email")
+  }
+
+  get password() {
+    return this.loginForm.get("password")
+  }
+
+  submitLoginForm(){
+    const user: IUser = {
+      email: this.email?.value!,
```

Comment:
(Gasp) is this a bang operator I see?

Reply:
Yes, that is a bang operator, reason being I know for a fact that these values are definitely going to be instantiated when the function is called. It is linked to a button click which is disabled unless all input fields have been entered and are valid so by the time you click on the button the values will exist

Reply:
Yes, I do love the fact that you disable your button if your form is invalid. Great use of directives!

However, never trust your DOM. Never trust your API. If they could definitely be instantiated they could never possibly be null and you wouldn't need a bang operator.

Add a guard clause to be safe.

Reply:
As per my reasoning above the values are definitely going to be instantiated before any operation is done with them so I don't think changes are required